crates/parser/src/grammar.rs

```diff
@@ -340,3 +342,29 @@ fn error_block(p: &mut Parser<'_>, message: &str) {
     p.eat(T!['}']);
     m.complete(p, ERROR);
 }
+
+fn delimited(
+    p: &mut Parser<'_>,
+    bra: SyntaxKind,
+    ket: SyntaxKind,
+    delim: SyntaxKind,
+    first_set: TokenSet,
+    mut parser: impl FnMut(&mut Parser<'_>) -> bool,
+) {
+    p.bump(bra);
+    while !p.at(ket) && !p.at(EOF) {
+        if !parser(p) {
+            break;
+        }
+        if !p.at(delim) {
+            if p.at_ts(first_set) {
+                p.error(format!("expected {:?}", delim));
```

Comment:
> [...] as this might get the parser stuck if I missed something

This is related to this right? Just a random unfinished  thought: The recovery here is basically unbounded right? Does it make sense to bound it to some hardcoded recovery limit?

Reply:
It assumes that the parser function consumes at least one token if it returns true, if it doesn't then this will lock up. All use sites should work this way though (will add a comment for that behavior)

Reply:
Yes I thought so :) However, my question was somewhat unrelated to that 😅. I was just wondering whether we shouldn't recover indefinitely. But on second thought that probably doesn't make any sense...

Again, unrelated to the above. IIuc, this

```rust
fn test() {
    hello(a <|>
    return 1
    ...
}
```

will pull the return expr into the arg list right? That's quite unfortunate.

Reply:
it will yes, but I don't think that's gonna be a problem, in this (broken syntax) spot, the only features that are relevant to the user are signature help and completions, neither of which would break from this (and assuming the return expression is termianted by a semicolon, nothing else after would be consumed either)